File: src/analyze.ts
/**
 * AI analysis of Creatio cases via the locally-installed Claude Code CLI.
 *
 * This module owns only the case-specific parts: the preset instructions, the
 * system prompt, and how case rows are flattened into text. The process
 * plumbing lives in ./shared/claudeRun.ts.
 *
 * INJECTION BOUNDARY: the `-p` argument is a fixed, app-authored instruction.
 * ALL untrusted content (case data + the user's free-text question) is written
 * to the child's stdin, never interpolated into argv or a shell string.
 *
 * Case text is raw third-party prose, so these runs use `cwd: "isolated"` with
 * no tools and no setting sources — the child can reason about the text but has
 * no filesystem to act on. See the injection-boundary comment in claudeRun.ts.
 */

import { runClaude, claudeAvailable, ClaudeCliError } from "./shared/claudeRun.js";
import type { CaseRow, CaseDetail, TimelineEntry } from "./caseLookup.js";

// Re-exported so existing importers (src/server.ts) keep working unchanged.
export { claudeAvailable, ClaudeCliError };

// ---------------------------------------------------------------------------
// Default model for analysis. Overridable per-run (CREATIO_APP_MODEL); set to
// a cheaper tier (e.g. "claude-sonnet-5", "claude-haiku-4-5") for faster runs.
// ---------------------------------------------------------------------------
export const DEFAULT_MODEL = "claude-opus-5";

// ---------------------------------------------------------------------------
// Preset instructions
// ---------------------------------------------------------------------------
export type Preset = "summarize" | "themes" | "actions" | "ask";

const PRESET_INSTRUCTIONS: Record<Exclude<Preset, "ask">, string> = {
  summarize:
    "Summarize and prioritize the Creatio support cases provided on stdin. " +
    "Give a 2-3 sentence overview, then a ranked list (most urgent first) of which " +
    "cases need attention and a short 'why' for each (age, who is blocking, severity).",
  themes:
    "Analyze the Creatio support cases provided on stdin and group them by common " +
    "theme or underlying root cause (e.g. report-card template bugs, setup/config " +
    "issues, data problems). For each theme list the case numbers and a one-line " +
    "explanation of the shared cause.",
  actions:
    "For each Creatio support case provided on stdin, state the single most useful " +
    "next action and who is currently blocking it (client vs support). Keep each " +
    "case to 1-2 lines, formatted as a list keyed by case number.",
};

const SYSTEM_PROMPT =
  "You are a concise support-operations analyst for a school-information-system (SIS) " +
  "team that customizes report-card templates in Creatio. You are given support-case " +
  "data (numbers, subjects, statuses, accounts, descriptions, and a merged timeline of " +
  "feed posts and emails). Analyze ONLY the data provided on stdin — do not invent facts " +
  "or use any tools. Note: timeline authors are unresolved, so never assert who wrote a " +
  "post; refer to content and @mentions only. Answer in clear, well-structured Markdown.";

// ---------------------------------------------------------------------------
// Context formatting
// ---------------------------------------------------------------------------
const MAX_CASES = 25; // keep token use / latency sane
const MAX_TIMELINE_PER_CASE = 12;
const MAX_TEXT = 1000; // chars per description / timeline entry

export interface AnalyzableCase extends CaseRow {
  detail?: CaseDetail;
}

function clip(s: string | undefined | null, n = MAX_TEXT): string {
  const t = (s || "").trim();
  return t.length > n ? t.slice(0, n) + " …[truncated]" : t;
}

function formatCase(c: AnalyzableCase, i: number): string {
  const lines: string[] = [];
  lines.push(`### Case ${i + 1}: ${c.Number} — ${c.Subject}`);
  lines.push(
    `Status: ${c.Status} | Account: ${c.Account || "?"} | Contact: ${c.Contact || "?"} | Created: ${c.CreatedOn}`
  );
  const d = c.detail || {};
  if (d.description) lines.push(`Description: ${clip(d.description)}`);
  const tl: TimelineEntry[] = d.timeline || (d.latest ? [d.latest] : []);
  if (tl.length) {
    const shown = tl.slice(-MAX_TIMELINE_PER_CASE);
    const omitted = tl.length - shown.length;
    lines.push(`Timeline (${tl.length} entr${tl.length === 1 ? "y" : "ies"}${omitted > 0 ? `, showing last ${shown.length}` : ""}):`);
    for (const e of shown) {
      const who = e.kind === "EMAIL" ? `${e.sender || "?"} → ${e.recipient || "?"}` : "feed post";
      lines.push(`- [${e.kind} ${e.ts}] (${who}) ${clip(e.text, 500)}`);
    }
  }
  return lines.join("\n");
}

export interface BuildContextResult {
  text: string;
  truncatedCases: number;
}

export function buildContext(cases: AnalyzableCase[]): BuildContextResult {
  const truncatedCases = Math.max(0, cases.length - MAX_CASES);
  const use = cases.slice(0, MAX_CASES);
  const blocks = use.map(formatCase);
  const header =
    `# ${use.length} Creatio support case${use.length === 1 ? "" : "s"}` +
    (truncatedCases > 0 ? ` (of ${cases.length}; ${truncatedCases} omitted to keep the analysis focused)` : "") +
    "\n";
  return { text: header + "\n" + blocks.join("\n\n"), truncatedCases };
}

// ---------------------------------------------------------------------------
// Runner
// ---------------------------------------------------------------------------
export interface AnalyzeOptions {
  preset: Preset;
  question?: string; // required when preset === 'ask'
  cases: AnalyzableCase[];
  model?: string; // optional override; defaults to DEFAULT_MODEL
  signal?: AbortSignal;
}

export interface AnalyzeCallbacks {
  onChunk: (text: string) => void;
  onDone: (meta: { costUsd?: number; totalTokens?: number; durationMs?: number; truncatedCases: number }) => void;
  onError: (err: ClaudeCliError) => void;
}

/**
 * Run an analysis, streaming text chunks via callbacks. Never rejects — all
 * outcomes are delivered through onDone / onError.
 */
export function analyzeCases(opts: AnalyzeOptions, cb: AnalyzeCallbacks): { kill: () => void } {
  const instruction =
    opts.preset === "ask"
      ? "Answer the user's QUESTION (below) using ONLY the Creatio support-case data on stdin."
      : PRESET_INSTRUCTIONS[opts.preset];

  const { text: context, truncatedCases } = buildContext(opts.cases);
  const question = (opts.question || "").trim();
  const stdin =
    (opts.preset === "ask" && question ? `QUESTION: ${question}\n\n` : "") + context;

  return runClaude(
    {
      instruction,
      systemPrompt: SYSTEM_PROMPT,
      stdin,
      // Untrusted prose in, no filesystem to act on: no tools, no repo context.
      cwd: "isolated",
      settingSources: [],
      model: opts.model || DEFAULT_MODEL,
      signal: opts.signal,
    },
    {
      onChunk: cb.onChunk,
      onDone: (meta) => cb.onDone({ ...meta, truncatedCases }),
      onError: cb.onError,
    }
  );
}

File: creatio_case_lookup/analyze.py
"""AI analysis of Creatio cases via the locally-installed Claude Code CLI.

This module owns only the case-specific parts: the preset instructions, the
system prompt, and how case rows are flattened into text. The process
plumbing lives in :mod:`creatio_case_lookup.claude_run`.

INJECTION BOUNDARY: the ``-p`` argument is a fixed, app-authored instruction.
ALL untrusted content (case data + the user's free-text question) is written
to the child's stdin, never interpolated into argv or a shell string.

Case text is raw third-party prose, so these runs use ``cwd="isolated"`` with
no tools and no setting sources — the child can reason about the text but has
no filesystem to act on. See the injection-boundary docstring in claude_run.py.
"""

from __future__ import annotations

import asyncio
import os
from typing import Any, Callable, Literal, TypedDict

from .claude_run import ClaudeCliError, RunHandle, RunSpec, claude_available, run_claude

# Re-exported so existing importers (server.py) keep working unchanged.
__all__ = [
    "DEFAULT_MODEL",
    "Preset",
    "PRESET_INSTRUCTIONS",
    "SYSTEM_PROMPT",
    "ASK_INSTRUCTION",
    "MAX_CASES",
    "MAX_TIMELINE_PER_CASE",
    "MAX_TEXT",
    "clip",
    "format_case",
    "build_context",
    "analyze_cases",
    "claude_available",
    "ClaudeCliError",
]

# ---------------------------------------------------------------------------
# Default model for analysis. Overridable per-run (CREATIO_APP_MODEL); set to
# a cheaper tier (e.g. "claude-sonnet-5", "claude-haiku-4-5") for faster runs.
# ---------------------------------------------------------------------------
DEFAULT_MODEL = "claude-opus-5"


def default_model() -> str:
    """The model for the app's runs: CREATIO_APP_MODEL (environment, then .env),
    else the `model` in this project's .claude/settings.json, else DEFAULT_MODEL.
    Passed as --model, which would otherwise override the settings file's."""
    from .claude_run import app_settings
    from .env import read_env_file

    return (
        os.environ.get("CREATIO_APP_MODEL")
        or (read_env_file().get("CREATIO_APP_MODEL") or "").strip()
        or app_settings().get("model")
        or DEFAULT_MODEL
    )

# ---------------------------------------------------------------------------
# Preset instructions
# ---------------------------------------------------------------------------
Preset = Literal["summarize", "themes", "actions", "ask"]

PRESET_INSTRUCTIONS: dict[str, str] = {
    "summarize": (
        "Summarize and prioritize the Creatio support cases provided on stdin. "
        "Give a 2-3 sentence overview, then a ranked list (most urgent first) of which "
        "cases need attention and a short 'why' for each (age, who is blocking, severity)."
    ),
    "themes": (
        "Analyze the Creatio support cases provided on stdin and group them by common "
        "theme or underlying root cause (e.g. report-card template bugs, setup/config "
        "issues, data problems). For each theme list the case numbers and a one-line "
        "explanation of the shared cause."
    ),
    "actions": (
        "For each Creatio support case provided on stdin, state the single most useful "
        "next action and who is currently blocking it (client vs support). Keep each "
        "case to 1-2 lines, formatted as a list keyed by case number."
    ),
}

ASK_INSTRUCTION = (
    "Answer the user's QUESTION (below) using ONLY the Creatio support-case data on stdin."
)

SYSTEM_PROMPT = (
    "You are a concise support-operations analyst for a school-information-system (SIS) "
    "team that customizes report-card templates in Creatio. You are given support-case "
    "data (numbers, subjects, statuses, accounts, descriptions, and a merged timeline of "
    "feed posts and emails). Analyze ONLY the data provided on stdin — do not invent facts "
    "or use any tools. Note: timeline authors are unresolved, so never assert who wrote a "
    "post; refer to content and @mentions only. Answer in clear, well-structured Markdown."
)

# ---------------------------------------------------------------------------
# Context formatting
# ---------------------------------------------------------------------------
MAX_CASES = 25  # keep token use / latency sane
MAX_TIMELINE_PER_CASE = 12
MAX_TEXT = 1000  # chars per description / timeline entry


class BuildContextResult(TypedDict):
    text: str
    truncatedCases: int


def _trim(s: str) -> str:
    # JS String.prototype.trim also strips U+FEFF; Python's strip() does not.
    return s.strip().strip("﻿").strip()


def clip(s: str | None, n: int = MAX_TEXT) -> str:
    t = _trim(s or "")
    return t[:n] + " …[truncated]" if len(t) > n else t


def _s(v: Any) -> str:
    """JS template-literal stringification for the fields we print."""
    if v is None:
        return "undefined"
    return str(v)


def format_case(c: dict, i: int) -> str:
    """`c` is a CaseRow dict (camel/Pascal keys as from Creatio) plus optional
    ``detail`` (a CaseDetail dict)."""
    lines: list[str] = []
    lines.append(f"### Case {i + 1}: {_s(c.get('Number'))} — {_s(c.get('Subject'))}")
    lines.append(
        f"Status: {_s(c.get('Status'))} | Account: {c.get('Account') or '?'} | "
        f"Contact: {c.get('Contact') or '?'} | Created: {_s(c.get('CreatedOn'))}"
    )
    d = c.get("detail") or {}
    if d.get("description"):
        lines.append(f"Description: {clip(d['description'])}")
    tl: list[dict] = d.get("timeline") or ([d["latest"]] if d.get("latest") else [])
    if tl:
        shown = tl[-MAX_TIMELINE_PER_CASE:]
        omitted = len(tl) - len(shown)
        lines.append(
            f"Timeline ({len(tl)} entr{'y' if len(tl) == 1 else 'ies'}"
            f"{f', showing last {len(shown)}' if omitted > 0 else ''}):"
        )
        for e in shown:
            who = (
                f"{e.get('sender') or '?'} → {e.get('recipient') or '?'}"
                if e.get("kind") == "EMAIL"
                else "feed post"
            )
            lines.append(f"- [{_s(e.get('kind'))} {_s(e.get('ts'))}] ({who}) {clip(e.get('text'), 500)}")
    return "\n".join(lines)


def build_context(cases: list[dict]) -> BuildContextResult:
    truncated_cases = max(0, len(cases) - MAX_CASES)
    use = cases[:MAX_CASES]
    blocks = [format_case(c, i) for i, c in enumerate(use)]
    header = (
        f"# {len(use)} Creatio support case{'' if len(use) == 1 else 's'}"
        + (
            f" (of {len(cases)}; {truncated_cases} omitted to keep the analysis focused)"
            if truncated_cases > 0
            else ""
        )
        + "\n"
    )
    return {"text": header + "\n" + "\n\n".join(blocks), "truncatedCases": truncated_cases}


# ---------------------------------------------------------------------------
# Runner
# ---------------------------------------------------------------------------
def analyze_cases(
    opts: dict,
    on_chunk: Callable[[str], None],
    on_done: Callable[[dict], None],
    on_error: Callable[[ClaudeCliError], None],
) -> RunHandle:
    """Run an analysis, streaming text chunks via callbacks. Never raises — all
    outcomes are delivered through on_done / on_error.

    ``opts`` keys (TS AnalyzeOptions): ``preset`` (required), ``question``
    (required when preset == "ask"), ``cases``, ``model`` (defaults to
    DEFAULT_MODEL), ``cancel`` (an ``asyncio.Event``; the TS ``signal``).
    ``on_done`` gets the run meta plus ``truncatedCases``.
    """
    preset = opts["preset"]
    instruction = ASK_INSTRUCTION if preset == "ask" else PRESET_INSTRUCTIONS[preset]

    ctx = build_context(opts.get("cases") or [])
    context, truncated_cases = ctx["text"], ctx["truncatedCases"]
    question = _trim(opts.get("question") or "")
    stdin = (f"QUESTION: {question}\n\n" if preset == "ask" and question else "") + context

    cancel: asyncio.Event | None = opts.get("cancel") or opts.get("signal")
    return run_claude(
        RunSpec(
            instruction=instruction,
            system_prompt=SYSTEM_PROMPT,
            stdin=stdin,
            # Untrusted prose in, no filesystem to act on: no tools, no repo context.
            cwd="isolated",
            setting_sources=[],
            model=opts.get("model") or default_model(),
            cancel=cancel,
        ),
        on_chunk=on_chunk,
        on_done=lambda meta: on_done({**meta, "truncatedCases": truncated_cases}),
        on_error=on_error,
    )
